### PR TITLE
Replace Collection with lightweight _SearchIndexer.

### LINKED ISSUE #664: Optimize `Collection` for internal use in `Project`
A possible optimization for signac 2.0 (or later) would be to reduce the amount of code we use in the `Collection` class when calling `Project.find_jobs`. See: https://github.com/glotzerlab/signac/blob/b29e0485c7998ba0c6d041e9ec15b533334d9b64/signac/contrib/project.py#L717

We copy a large amount of data used from the `Project`'s internal caches during the construction of a `Collection` and calling its `find` method. At the very least, we never use the `Collection`'s ability to interact with data on disk or its ability to automatically generate ids (primary keys) for new records in the context of a signac `Project`, so we could eliminate some logic there in a cut-down class.

_Originally posted by @bdice in https://github.com/glotzerlab/signac/issues/652#issuecomment-1002303783_

I attempted an optimization in 13f2a8fb205f65d49421f4f4009ee3d78d00f9bf but it was [unclear if copy/reference semantics would be correct in the resulting indices](https://github.com/glotzerlab/signac/pull/652#discussion_r777170033) outside the context of a signac `Project`'s limited usage of `Collection`. A smaller class that is designed for the actual use case of signac's `Project.find_jobs` could act as an internal cache with only the necessary logic (e.g. no file I/O or id generation).

### PR DESCRIPTION
**Pull request by bdice**
## Description
This PR (draft) minimizes the Collection class for the two use cases it has in signac's internals: detecting schema and finding job ids.

The `Collection` class is 1174 lines and the `_SearchIndexer` class is 273 lines (including a helper method that I inlined, which makes it appear larger).

### To-do:
- [x] Measure performance
  - See benchmarks below. This PR gives a 20-30% speedup for small projects and 7-15% speedup for large projects.
- [x] Change index format from list of dicts with an `_id` key like `[{"a": 0, "_id": "abcdef"}]` to a dict like `{"abcdef": {"a": 0}}`. This simplifies several areas of the code and makes `_SearchIndexer`'s semantics dict-like.
- [x] Expand schema/find tests to cover all filter use cases
- [x] Reduce `_SearchIndexer` class to bare minimum amount of code
- [x] 100% test coverage of `_index.py`
- [ ] Evaluate whether to remove `_SearchIndexer` and use a plain `dict` with helper functions?
  - This is viable. It would mean removing the public methods `find` and `build_index`, and instead defining them as free functions that act on a `dict` with certain assumptions about structure. Do we like this idea?
- [ ] Remove `Collection` from public API
  - Planning a follow-up PR.
- [ ] Consider moving "filter" docs to a better place
  - Investigate this in a follow-up PR when removing `Collection`.

## Motivation and Context
Resolves #664.

## Types of Changes
- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt) and added all related issue and pull request numbers for future reference (if applicable). See example below.

### CODE REVIEW

**Comment by bdice** [comment]
Benchmarks below! I only compared the "find/search" benchmarks below -- all the other benchmarks are basically unchanged between master, next, and this branch.

Overall I would say the time spent searching appears to have dropped by 20-30% for lean search filters over small projects, and by 7-15% for richer search filters over larger projects. Not bad.

|               Commit              | master (dac370a0) | next (d26930fa) | optimize-collection (37b84f1b) |
|:---------------------------------:|-------------------|-----------------|--------------------------------|
| time_search_lean_filter (100)     | 351 us            | 385 us          | 265 us                         |
| time_search_lean_filter (1,000)   | 5.82 ms           | 6.01 ms         | 4.10 ms                        |
| time_search_lean_filter (100,000) | 156 ms            | 156 ms          | 144 ms                         |
| time_search_rich_filter (100)     | 1.43 ms           | 1.50 ms         | 1.24 ms                        |
| time_search_rich_filter (1,000)   | 25.7 ms           | 25.8 ms         | 23.5 ms                        |
| time_search_rich_filter (100,000) | 281 ms            | 278 ms          | 259 ms                         |

**Review by csadorf — CHANGES_REQUESTED** [review]
Thanks for taking this on!

Since this is largely a copy of the full `Collection` class, is the plan to remove it in a subsequent step or are we keeping both? If we keep both I think we should make an attempt at using composition to avoid the code duplication.

**Comment by bdice** [comment]
> is the plan to remove it in a subsequent step or are we keeping both? If we keep both I think we should make an attempt at using composition to avoid the code duplication.

As I mentioned in our dev meeting today, my plan is to remove the `Collection` class entirely in a follow-up PR. Because I changed the class's internal data structures, it would probably be awkward to try and retain the old class while using composition or inheritance to reduce duplication.

File: signac/contrib/_slim_collection.py
Comment (bdice) [reply signac/contrib/_slim_collection.py:27]: @csadorf I appreciate these suggestions! I had a similar series of ideas -- I kept `_SlimCollection` for the initial round review partly because I knew the name should change but didn't want to commit to a name until reviewing its use cases and getting feedback like what you provided. I lean towards `_Index`, which should not conflict with any other internals now that we've removed other forms of indexing from the `next` branch. My only hesitation is that it has a method `build_index`, which is an index organized by key. I wonder if that method (or the class) should be named differently. Otherwise the result of `_Index.build_index` is just a `dict` with a particular structure, while it sounds like it should be a class factory method.

File: signac/contrib/_slim_collection.py
Comment (bdice) [reply signac/contrib/_slim_collection.py:269]: Agreed. I'll remove this.

File: signac/contrib/_slim_collection.py
Comment (bdice) [reply signac/contrib/_slim_collection.py:271]: Correct. I had wondered about adding an early-exit for `$not` and `$and` after reducing the results (the `$or` case is at the end, so exiting early is meaningless). I have now added it.

**Comment by vyasr** [comment]
I will get around to reviewing this soon, but in the meantime @bdice do you still think

> Evaluate whether to remove _Index and use a plain dict with helper functions? 

is viable and worth pursuing? What would be the pros and cons of doing that vs the current approach? I'm all for using the most stripped down solution that we can in 2.0. The other outstanding items in the PR description's checklist can definitely be addressed in a follow-up PR, but if there's a chance that we're going to end up removing this new class again because there's a simpler alternative it would be good to figure that out now.

**Comment by bdice** [comment]
@vyasr dict-only with helper methods is viable. “Worth pursuing” is up for debate. Pro: no class = simpler? Con: pure dict inputs must have a particular structure that has no centralized home for documentation without a class.

**Comment by csadorf** [comment]
> There is no internal state besides the data in the index.

But would those indexes not be the best argument for a dedicated class to maintain that state? Building those indexes are the primary purpose of the class, no? Also, keep in mind that the `Collection` class was originally supposed to be a near drop-in for `pymongo.Collection`, that would certainly be lost at that point.

**Comment by bdice** [comment]
> But would those indexes not be the best argument for a dedicated class to maintain that state?

The `Collection` class had support for building indices and tracking those internally in the class state. However, signac's use of the `Collection` was limited to one-off cases where the `Collection` was constructed, searched, and then went out of scope. As a result, the cached indices of the `Collection` class were _never used more than once_, and thus it made no sense to cache them. The [logic for determining whether the index was dirty and needed to be rebuilt](https://github.com/glotzerlab/signac/blob/5111cde0d36eb8ee230a5c030e33fc1ae2292859/signac/contrib/collection.py#L579) was unnecessary because the index was only queried once, and thus it was always "dirty" before that first query.

For simplicity, I removed key-based indices from the internal state of the `_Index` object. The `_build_job_statepoint_index` function has been refactored in this PR to store key-based indices locally in that function, rather than in the internal state of the `_Index` object: https://github.com/glotzerlab/signac/blob/047333fad7724937b5c933166a3ec570c81d2ac2/signac/contrib/schema.py#L88-L93

File: signac/contrib/_slim_collection.py
Comment (bdice) [reply signac/contrib/_slim_collection.py]: For now, I've renamed this class to `_Index`.

**Comment by bdice** [comment]
@csadorf @vyasr This is ready for review -- I would like to keep the current PR scope and hold off on switching to a pure `dict` with free functions for the build/find features. If you think that using a plain `dict` (rather than a subclass) is an improvement, please say so, but I would like to do that in a follow-up PR rather than this one.

File: signac/contrib/_index.py
Comment (vyasr) [inline signac/contrib/_index.py:82]: AFAICT this function is almost exclusively called internally except for a single usage in schema code. Could we change the input to be a list of strings already split? That would remove an unnecessary join and resplit that occurs when this is called from `_find_expressions`.

File: signac/contrib/schema.py
Comment (vyasr) [inline signac/contrib/schema.py:52]: I don't think this should be defined here. It belongs with _Index (or Collection until that class is removed).

File: signac/contrib/_slim_collection.py
Comment (vyasr) [reply signac/contrib/_slim_collection.py]: I would strongly prefer a more disambiguated name. While I agree that `_Index` is not currently conflicting with anything, I don't want to hobble our ability to use related names in the future. Otherwise we could run into the same issues that we now have in signac-flow where vague naming leads to ambiguity between concepts like groups, aggregates, bundles, etc. Having reviewed the code, since the critical operation is the ability to search the elements maybe it's something like a `_SearchableIndex`? I would definitely like to avoid the term `Collection` given a) the overloads associated with collections.abc (and now SyncedCollections) and b) we are no longer trying to specifically make this class behave so much like a MongoDB collection

File: signac/contrib/_index.py
Comment (bdice) [reply signac/contrib/_index.py:55]: This constructor isn't accurate anymore, I have pushed changes that make this suggestion obsolete.

File: signac/contrib/_index.py
Comment (bdice) [reply signac/contrib/_index.py]: Yeah, I agree. I did this a little bit. I probably need to improve the docs on restrictions, e.g. no dots in keys.

Comment (bdice) [reply signac/contrib/_index.py]: Added restrictions (data must be JSON-encodable and dots are not allowed in keys).

File: signac/contrib/_index.py
Comment (bdice) [reply signac/contrib/_index.py:88]: I agree but I would prefer to do this across our entire codebase in a single PR. We are currently consistent about using `_id` everywhere.

Comment (bdice) [reply signac/contrib/_index.py:88]: https://github.com/glotzerlab/signac/issues/679

File: signac/contrib/_index.py
Comment (bdice) [reply signac/contrib/_index.py:299]: This encode/decode blocks a large class of invalid inputs. All valid queries are JSON-encodable dicts, and this ensures we don't have arbitrary Python objects being passed in to queries. We have a second validation step in `_valid_filter` that ensures we block cases like `[]`. However, I think that helper function could be eliminated: it appears that the only invalid filter is a top-level list, `[]`. Doing a round-trip through JSON encoding and ensuring that the filter is a `dict` should be enough for validation.

File: signac/contrib/schema.py
Comment (bdice) [reply signac/contrib/schema.py:52]: Good idea! I did that. For now I moved to `_index.py` since `collection.py` will eventually be deleted.

File: signac/contrib/_index.py
Comment (bdice) [reply signac/contrib/_index.py:101]: Hmm. I was able to eliminate one join by re-using the input key in the error check instead of computing `nodes.join(".")` and I made some other related improvements. See f4a05d415c3ffd55f18bfeaa9d3711b478a34dae. However, I'm not sure if we want to pass in `nodes` directly... I think that is more awkward and I don't see the performance benefit. We have to check if the original dotted key is present in the document and raise if so. Also I don't think it's worth changing the API. Maybe we can do that in a follow-up PR if you're still sure this would be helpful?

This microbenchmark in particular is kind of interesting: https://github.com/bdice/python_microbenchmarks/blob/main/microbenchmarks/nested_access.py I see that `recurse_all` is about 30% faster than `one_direct_access` when dealing with 2 levels of nesting, and about 40% faster when dealing with no nesting, so I changed our access pattern.

**Comment by bdice** [comment]
@csadorf @vyasr This is ready for you to take another look, there are multiple open comment threads.

I did some refactoring from @vyasr's review comments and was able to get some additional speedups. The post above with benchmarks has been updated. https://github.com/glotzerlab/signac/pull/667#issuecomment-1032214136

File: signac/contrib/_index.py
Comment (vyasr) [reply signac/contrib/_index.py:101]: My goal was mainly to avoid the redundant work. I don't feel strongly that we need to change the API as long as we can minimize the number of redundant string operations there.

That is interesting. It makes sense in hindsight, but I wouldn't have thought about it beforehand.

File: signac/contrib/_index.py
Comment (vyasr) [inline signac/contrib/_index.py:164]: To be clear, this is the kind of operation that I was hoping to avoid when I suggested changing the build_index API. We rejoin the string just to split it again immediately inside build_index. That sequence of operations isn't exactly super expensive, but it's also not free, especially given that we see the impact of microoptimizations on the level `isinstance` vs `type` checks:
```
In [20]: %timeit '.'.join('a.b.c'.split('.')[1:]).split('.')
263 ns ± 22.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each
```

File: signac/contrib/_index.py
Comment (vyasr) [inline signac/contrib/_index.py:312]: I think it's already always a `set`, right? In case there's a code path that I'm missing, can we change that?
```suggestion
        return self._find_result(filter_)
```

File: signac/contrib/_index.py
Comment (vyasr) [inline signac/contrib/_index.py:310]: Same question as before regarding `isinstance` vs `type`.

**Review by vyasr — CHANGES_REQUESTED** [review]
Next round of suggestions. I'm really getting into the nitty gritty of some of these microoptimizations, but I think it's worth spending the effort now since it's highly likely that once this PR is merged we probably won't touch this code again for a long time.

File: signac/contrib/_index.py
Comment (vyasr) [reply signac/contrib/_index.py]: Right, but at this stage haven't we always already parsed the filter using `filterparse.parse_filter` to validate the inputs? I'm pretty sure that since we are no longer exposing this API publicly we can make more assumptions about the validity of the inputs than we are currently making. And JSON encoding is quite expensive.

Comment (vyasr) [reply signac/contrib/_index.py]: Here's my quick analysis. We call `find` in schema.py::_build_job_statepoint_index with no arguments. That gets used in `import_export.py` and in `Project.detect_schema`. We also call `_Index.find` in `Project._find_job_ids`, where the filter has already been parsed by `parse_filter`. I think that's all the calls to this function that exist. So I think we don't need this extra normalization at all.

**Comment by bdice** [comment]
@vyasr That was an excellent round of code review, thanks for the attention to detail. I like your suggestions and will work through them soon!

File: signac/contrib/_index.py
Comment (csadorf) [reply signac/contrib/_index.py]: Yes, the purpose is input validation and normalization and it certainly fixed related bugs when first introduced. I would recommend that in addition to @vyasr ' analysis, we ensure that this is covered by unit tests for unsanitized input before removing or replacing it.

**Comment by csadorf** [comment]
> Next round of suggestions. I'm really getting into the nitty gritty of some of these microoptimizations, but I think it's worth spending the effort now since it's highly likely that once this PR is merged we probably won't touch this code again for a long time.

Should we do the micro optimizations in a separate PR to avoid this getting too bloated?

File: signac/contrib/_index.py
Comment (csadorf) [reply signac/contrib/_index.py:50]: Considering that this class is primarily used internally and there is apparently no expectations of maintaining indexes more than once, I would be open to this change, but like you said, outside of scope for this PR.

File: signac/contrib/import_export.py
Comment (csadorf) [inline signac/contrib/import_export.py:79]: One could use `defaultdict` instead.

File: signac/contrib/_index.py
Comment (csadorf) [inline signac/contrib/_index.py:31]:
```suggestion
    d : dict
```

**Review by csadorf — APPROVED** [review]
No concerns, looks good to me! Nice work! 😃 

No objections if you want to work through @vyasr ' suggestions before merge, but you could also consider to split these up.

**Comment by vyasr** [comment]
I'm not concerned about most of my suggestions expanding the PR scope, they should almost entirely be one or two line changes to existing code paths. The only one that might be worth pushing to a subsequent PR is the changes to the JSON-based normalization. @bdice if you want to wait to add tests and see if removing that is safe, I'm fine with that. It's up to you though, I can also review those changes in this PR if you choose to add them here.

File: signac/contrib/_index.py
Comment (bdice) [reply signac/contrib/_index.py:166]: This split occurs `O(number of keys in a filter)` times. Most of the other microoptimizations happen `O(jobs)` times, which are the cases I care about most. I am going to just resolve this and say it's good enough for now.

File: signac/contrib/_index.py
Comment (bdice) [reply signac/contrib/_index.py]: Nice find. I removed that helper function and the extraneous copy.

**Comment by bdice** [comment]
@vyasr I think I've addressed all your suggestions. Please review when you're able, or merge this if you think it's ready.

File: signac/contrib/import_export.py
Comment (csadorf) [reply signac/contrib/import_export.py:79]: That's fair.

File: signac/contrib/_index.py
Comment (vyasr) [reply signac/contrib/_index.py:96]: Yes, that is my recollection as well and I'm comfortable working under that assumption.

File: signac/contrib/_index.py
Comment (vyasr) [inline signac/contrib/_index.py:4]:
```suggestion
"""Implement class for indexing signac Projects."""
```

File: signac/contrib/_index.py
Comment (vyasr) [inline signac/contrib/_index.py:98]:
```suggestion
                # `isinstance(instance, cls)` is typically faster than `type(instance) is cls`
                # when the answer is True, but it is slower when it is False. Since we
                # expect lists and dicts to occur infrequently here, we optimize for the
                # False path using the `type` based check.
                if type(v) is list:
                    index[_to_hashable(v)].add(_id)
                elif type(v) is dict:
```

File: signac/contrib/_index.py
Comment (vyasr) [inline signac/contrib/_index.py:103]:
```suggestion
            # Raise an exception if the original key is present and has dots.
```

File: signac/contrib/_index.py
Comment (vyasr) [inline signac/contrib/_index.py:120]:
```suggestion
        value
```

**Review by vyasr — APPROVED** [review]
Great work here, LGTM! I have some minor suggestions for comments that I'll commit and then merge.

**Review by vyasr — CHANGES_REQUESTED** [review]
Ack sorry I just realized we missed two things:
1. Choosing a less ambiguous name: https://github.com/glotzerlab/signac/pull/667#discussion_r801310592
2. Analyzing the json logic currently used to sanitize filters: https://github.com/glotzerlab/signac/pull/667#discussion_r805197820

Let's push 2 to future work so that we don't increase the scope of this PR any further, and as @csadorf pointed out we should add more tests etc before making any change there. However, I do think we do need to address 1 before we merge this PR since I am quite wary of introducing another generically named object into our namespace. Some options:
1. `_SearchableIndex`
2. `_SearchableDict`
3. `_SearchableKeyStore`

Other thoughts? As I stated previously, I would like to avoid the `Collection` terminology to minimize collisions with `collections.abc.Collection`, especially since we are no longer trying to mimic MongoDB functionality here. I'm open to other suggestions though.

**Comment by csadorf** [comment]
> However, I do think we do need to address 1 before we merge this PR since I am quite wary of introducing another generically named object into our namespace. Some options:
> 
> 1. `_SearchableIndex`
> 2. `_SearchableDict`
> 3. `_SearchableKeyStore`
> 
> Other thoughts? As I stated previously, I would like to avoid the `Collection` terminology to minimize collisions with `collections.abc.Collection`, especially since we are no longer trying to mimic MongoDB functionality here. I'm open to other suggestions though.

I just looked at the doc-string for the class it says “A searchable collection of dicts.” which seems pretty accurate.
1. I think that the name `_SearchableDict` is not appropriate, because it implies that we have a single dict that can be searched, but here we have actually a “collection” of dicts that can be indexed and searched. I understand that is technically a single dict linking these individual dicts by their primary keys, but that’s the wrong focus IMO.
1. I agree that the term “index” does not be in the name since an index is a specific data structure that speeds up the search process and within the context of this class constitutes an implementation detail. Also, there could be multiple indexes for one “searchable collection of dicts”.
1. After going through a bunch of taxonomies for databases (https://raima.com/database-terminology/, https://www.lifewire.com/databases-glossary-1019603#toc-index) , how would you feel about simply calling it `_Database`? We could be more specific, but in the end it’s the primary “signac database” structure, right?

**Comment by bdice** [comment]
@csadorf I think the definition you posed for “Index” explains what this class does:

> an index is a specific data structure that speeds up the search process

Meanwhile, I do not think a database is the right word for this — a database has create/read/update/delete operations and is typically (though not always) persistent. This is a construct-once, read-only, non-persistent class. The database in signac is the `Project` class.


I would prefer `_SearchableIndex`.

**Review by vyasr — APPROVED** [review]
I don't want to waste too much time on this either, I just also didn't want to merge and completely forget this either. `_SearchIndexer` is fine with me. We can see where we end up on the dict refactor and revisit this discussion if the class ends up sticking around.

**Pull request merged**
